### ISSUE TITLE
Support for scan-build (clang static analyzer)

### ISSUE THREAD

**Issue by Harvie** [mongoose-os/mos#39]
Hello,
can you please add support for `scan-build` in the mongoose os build system?

https://clang-analyzer.llvm.org/scan-build.html

It is easy to use. You just add `clang` package to docker image and use `scan-build` to invoke whatever build system you are already using. Eg. instead of `make` you invoke `scan-build make`. It will wrap GCC invokations to do clang static analysis as well. It offers interresting warnings and insight into possible issues present in the code, which are otherwise not detected by compiler.

Note that while scan-build is part of clang compiler, it will still use gcc compiler to do the actual compiling unless you configure your system to use clang (which should work as drop-in gcc replacement in case you want to try it as well).

Besides of compiling the code in gcc as usual, this will produce additional warnings printed in the console and will also produce directory containing verbose build protocol to be later examined by user. I think this would allow us to write more robust code.

Unfortunately i cannot simply invoke it as `scan-build mos build --local`, because the wrapper will obviously not affect build process inside docker container. Only way to make this work is adding the support directly to the docker image.

Perhaps this can be implemented as optional mode for mos, eg. activated as: `mos build --local --scan-build`

**Comment by rojer** [comment]
i love static analysis tools, the more we can run on the code the better.
i will take a look at this when i have the time, or if you can do it yourself and put up PRs, that would be even better,

**Comment by Harvie** [comment]
> i will take a look at this when i have the time, or if you can do it yourself and put up PRs, that would be even better

I might, but TBH i don't even know where to start... Where are the sources of docker image?

**Comment by rojer** [comment]
https://github.com/cesanta/mongoose-os/tree/master/tools/docker

**Comment by Harvie** [comment]
> https://github.com/cesanta/mongoose-os/tree/master/tools/docker

But how do i force mos tool on my machine to use some unofficial development version of docker image?

I am not tremendously experienced with docker, but if i remember correctly it might even be possible to add some kind of "overlay image" which would stack the scan-build layer on top of the mongoose image without need to modify it. But modifiing the image directly might be easier...

**Comment by rojer** [comment]
it's simple, actually: you build local image under a different tag and change deps/modules/mongoose-os/platforms/PLATFORM/sdk.version to use it.

so, since we are talking ubuntu image, you:
1) modify [Dockerfile](https://github.com/cesanta/mongoose-os/blob/master/tools/docker/ubuntu/Dockerfile-ubuntu-build)
2) `make docker-build TAG=custom`
3) edit [sdk.version](https://github.com/cesanta/mongoose-os/blob/master/platforms/ubuntu/sdk.version) to use `docker.io/mgos/ubuntu-build:custom`
4) build your app

then iterate steps 1, 2, 4

**Comment by Harvie** [comment]
> https://github.com/cesanta/mongoose-os/blob/master/tools/docker/ubuntu/Dockerfile-ubuntu-build

There already seems to be clang and even valgrind already included in the image...
However valgrind does dynamic analysis, so correct me if i am wrong, but it cannot be used to analyze ESP32 code in docker on PC...

**Comment by rojer** [comment]
yeah, dynamic analysis on the target is not possible. the best we can do is analyze app built for ubuntu platform.

**Comment by Harvie** [comment]
> the best we can do is analyze app built for ubuntu platform.

What is this ubuntu platform for? is it just for testing or for actual devices? Unfortunately i use mos modules which do not have "ubuntu" listed as platform in mos.yml. Also my code often starts allocating memory only in case it gets some data from hardware peripherals which are not present on ubuntu, so i guess this will not be very helpful in my case...

**Comment by Harvie** [comment]
Also when i inspect the dockerfile https://github.com/cesanta/mongoose-os/blob/master/tools/docker/ubuntu/Dockerfile-ubuntu-build

It seems to describe what packages are installed in the image, but there is no mention of any buildscripts being invoked... I've looked at mgos dockerhub account and it seems there are lots of images. eg. this:

https://hub.docker.com/layers/mgos/fwbuild-instance/2.18.0/images/sha256-4a49a533f6a256f0296ad7bf5f1e112c3323f141ae824cab0b0121d8ccd81630?context=explore

seems to specifiy `ENTRYPOINT ["/usr/local/bin/fwbuild-instance"]` which is probably the script i am looking for. However i can't find it anywhere on github. I would like to find where the actual building is done, so i can wrap that command with scan-build.

**Comment by rojer** [comment]
build process itself is driven by the make file: https://github.com/cesanta/mongoose-os/blob/master/platforms/ubuntu/Makefile.build

mos runs `make -f Makefile.build` with a bunch of variables.

```
docker run --name mos_build_2020-12-05T17-20-14-00_5434562690696808693 --rm -i -v ... a lot of volume mappings ... make '-j' '8' '-C' '/app' '-f' '/home/rojer/allterco/shelly-homekit/deps/modules/mongoose-os/platforms/ubuntu/Makefile.build' 'all' ... while bunch of variables
```


run mos build --verbose and find the line that starts with "Docker arguments:" for full invocation

**Comment by Harvie** [comment]
So perhaps it might be enough to replace
`docker run ... make '-j' '8' '-C' '/app'` with
`docker run ... scan-build make '-j' '8' '-C' '/app'`

that way we would only need to add the clang-tools package

**Comment by Harvie** [comment]
But when i copy all the arguments printed in mos build --verbose and try to run it using docker command, it just prints `0` and exits.

**Comment by Harvie** [comment]
Also i can't build "custom" tag, because it looks for the tag named like that in other repositories...

```
Step 11/22 : RUN git clone --recursive --branch $DOCKER_TAG https://github.com/mongoose-os/esp-idf /opt/Espressif/esp-idf
 ---> Running in 17dbc7e5fa50
Cloning into '/opt/Espressif/esp-idf'...
fatal: Remote branch custom not found in upstream origin
The command '/bin/sh -c git clone --recursive --branch $DOCKER_TAG https://github.com/mongoose-os/esp-idf /opt/Espressif/esp-idf' returned a non-zero code: 128
make: *** [../docker.mk:24: docker-build-esp32-build] Chyba 128
```

**Comment by Harvie** [comment]
this also failed: `make docker-build`, because there is no `latest` in other repos (same way as there is no `custom` tag)
but i was able to build like this: `make docker-build DOCKER_TAG=3.3-r6`.

and i was able to run scan-build with modified mos tool as described in https://github.com/mongoose-os/mos/issues/39#issuecomment-739324272

BUT it seems that scan-build does call different GCC than the one needed in esp-idf, because it fails with following error:
`gcc: error: unrecognized command line option '-mlongcalls'`

So we probably need to figure this out...

**Comment by rojer** [comment]
> Also i can't build "custom" tag, because it looks for the tag named like that in other repositories...

forget about ESP32, it won't work. focus on ubuntu platform for using such tools. ubuntu platform is exactly for that - we don't expect you to deploy apps running on it, but the aim is to make it mostly work, to be able to run a good chunk of your logic under ASAN, UBSAN, valgrind and such.

**Comment by rojer** [comment]
for example, i do most of my logic developement of the shelly-homekit firmware under Ubuntu - the ShellyU target [here](https://github.com/mongoose-os-apps/shelly-homekit/blob/master/mos.yml#L456).

**Comment by Harvie** [comment]
> forget about ESP32, it won't work

I think this one might actualy be doable as it's static analysis. it just goes through the sources and then calls actual compiler, which might be the gcc from esp32 toolchain.

**Comment by Harvie** [comment]
scan-build has these options, perhaps we need to set them properly: `[--use-c++ [=compiler_path]] [--use-cc [=compiler_path]] `

**Comment by Harvie** [comment]
I guess i am giving up for now :-(

> for example, i do most of my logic developement of the shelly-homekit firmware under Ubuntu

But when i do `mos build --local --platform=ubuntu` the build fails basicaly on any module which contains `build_vars`

**Comment by rojer** [comment]
if it's a pre-compilation step anyway, all you need to do is just amke your project _build_ for --platform=ubuntu, not necessarily run.
this can be usually be achieved by removing some platform restrictions in manifests and adding a bunch of stubs.

**Comment by Harvie** [comment]
I guess the ubuntu build might be the way as long as i am able to build whole codebase.

Also i recently started running out of disk space and i can't fit both docker images to my laptop :-D These are relatively big, so i would preffer to only have one...

**Comment by rojer** [comment]
oh man :) get yourself a nice, big SSD :)

**Comment by Harvie** [comment]
> oh man :) get yourself a nice, big SSD :)

haha. i already have SSD, but i am considering laptop upgrade in next year, so i don't see point in buying new SATA drive, when my next laptop is unlikely to use SATA...

**Comment by Harvie** [comment]
Anyway... i figured out another two options for analyzing directly during xtensa build...

1.) Get scan-build support directly by esp-idf: https://github.com/espressif/esp-idf/issues/6213
2.) Upgrade to gcc 10 which has recently integrated static analyzer as wel:

https://developers.redhat.com/blog/2020/03/26/static-analysis-in-gcc-10/
https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html

**Comment by Harvie** [comment]
ESP-IDF just added support for GCC 11: https://github.com/espressif/esp-idf/issues/6214
Which contains static analyzer as well. However it is considered to be complementary to stuff that other analyzers cannot do. So it still makes sense to use both GCC11 -fanalyzer and clang-scan.

**Comment by Harvie** [comment]
Also ESP-IDF just added option to do statical analysis using clang: https://github.com/espressif/esp-idf/issues/6213
